qemu/tests/block_hotplug.py

```diff
@@ -88,153 +76,81 @@ def unplug_device(vm, get_disk_cmd, device):
-                                  logging.info)
-            if params.get("disk_op_cmd"):
-                if params.get("os_type") == "linux":
-                    test_cmd = params.get("disk_op_cmd") % (disk, disk)
-                elif params.get("os_type") == "windows":
-                    test_cmd = params.get("disk_op_cmd") % (disk_letter[num],
-                                                            disk_letter[num])
-                    test_cmd = utils_misc.set_winutils_letter(session, test_cmd)
-                else:
-                    test.cancel("Unsupported OS type '%s'" % params.get("os_type"))
-
-                status, output = session.cmd_status_output(test_cmd,
-                                                           timeout=disk_op_timeout)
-                if status:
-                    test.fail("Check for block device failed "
-                              "after hotplug, Output: %r" % output)
-            session.close()
-
-        sub_type = params.get("sub_type_after_plug")
-        if sub_type:
-            error_context.context(context_msg % (sub_type, "after hotplug"),
-                                  logging.info)
-            utils_test.run_virt_sub_test(test, params, env, sub_type)
-            if sub_type == "shutdown" and vm.is_dead():
-                return
-
-        sub_type = params.get("sub_type_before_unplug")
-        if sub_type:
-            error_context.context(context_msg % (sub_type, "before unplug"),
-                                  logging.info)
-            utils_test.run_virt_sub_test(test, params, env, sub_type)
+        plug_tag = "before_unplug"
+        run_sub_test(params, plug_tag)
 
         error_context.context("Unplug block device (iteration %d)" % iteration,
                               logging.info)
-        for controller in controller_list:
-            controller_id = controller.get_param("id")
-            for device in controller_device_dict.get(controller_id, []):
-                unplug_status = unplug_device(vm, get_disk_cmd, device)
-                if not unplug_status:
-                    test.fail("Failed to unplug disks '%s'" % device.get_param("id"))
-                device_list.remove(device)
-            controller.unplug(vm.monitor)
-        for device in device_list:
-            unplug_status = unplug_device(vm, get_disk_cmd, device)
-            if not unplug_status:
-                test.fail("Failed to unplug disks '%s'" % device.get_param("id"))
-
-        sub_type = params.get("sub_type_after_unplug")
-        if sub_type:
-            error_context.context(context_msg % (sub_type, "after unplug"),
-                                  logging.info)
-            utils_test.run_virt_sub_test(test, params, env, sub_type)
+        disks_before_unplug = find_disk(vm, get_disk_cmd)
+        for device in reversed(device_list):
+            vm.devices.simple_unplug(device, vm.monitor)
```

Comment:
Btw the `simple_unplug` allows the same treatment as `simple_hotplug`. You can check for `False` and fail the test. This is actually even more important than `hotplug` because the disk can be unplugged successfully while unplugging of the HBA (controller) might fail. Currently in that case this test would passed but it should fail instead.